File: src/python/day10.py
import z3
from collections import deque

example_input = """
[.##.] (3) (1,3) (2) (2,3) (0,2) (0,1) {3,5,4,7}
[...#.] (0,2,3,4) (2,3) (0,4) (0,1,2) (1,2,3,4) {7,5,12,7,2}
[.###.#] (0,1,2,3,4) (0,3,4) (0,1,2,4,5) (1,2) {10,11,11,5,10,5}
"""

def parse_input(input) -> list[list[str], list[list[int]], list[int]]:
  res = []
  for line in input.strip().split("\n"):
    indicator_light_diagram = [ char == '#' for char in line.split(" ")[0]][1:-1]
    button_wiring_schematics = [ list(map(int, button[1:-1].split(","))) for button in line.split(" ")[1:-1]]
    joltage_requirements = list(map(int, line.split(" ")[-1][1:-1].split(",")))
    res.append((indicator_light_diagram, button_wiring_schematics, joltage_requirements))
  return res

def solve_with_indicator_lights(machine):
  indicator_light_diagram, button_wiring_schematics, joltage_requirements = machine

  solver = z3.Optimize()
  presses = [ z3.Int(f"button_{i}") for i in range(len(button_wiring_schematics)) ]

  # Button presses are positive integers
  for i in range(len(button_wiring_schematics)):
    solver.add(presses[i] >= 0)

  # The indicator light is on if the sum of the number of presses for each button that affects the indicator light is odd
  for i in range(len(indicator_light_diagram)):
    terms = []
    for j in range(len(button_wiring_schematics)):
      if i in button_wiring_schematics[j]:
        # if the button affects the indicator light, add the number of presses to the terms
        terms.append(presses[j])
    equation = z3.Sum(terms) % 2 == int(indicator_light_diagram[i])
    solver.add(equation)

  # Minimize the total number of button presses
  solver.minimize(z3.Sum(presses))

  if solver.check() == z3.sat:
    return solver.model().eval(z3.Sum(presses)).as_long()
  return -1

def part1(input):
  total_presses = 0
  for machine in input:
    presses = solve_with_indicator_lights(machine)
    total_presses += presses
  return total_presses

def solve_with_joltage(machine):
  indicator_light_diagram, button_wiring_schematics, joltage_requirements = machine

  solver = z3.Optimize()
  presses = [ z3.Int(f"button_{i}") for i in range(len(button_wiring_schematics)) ]
  
  # Button presses are positive integers
  for i in range(len(button_wiring_schematics)):
    solver.add(presses[i] >= 0)

  # The sum of the number of presses for each button that affects the joltage requirement must equal the joltage requirement
  for i in range(len(joltage_requirements)):
    terms = []
    for j in range(len(button_wiring_schematics)):
      if i in button_wiring_schematics[j]:
        # if the button affects the joltage requirement, add the number of presses to the terms
        terms.append(presses[j])
    equation = z3.Sum(terms) == joltage_requirements[i]
    solver.add(equation)

  # Minimize the total number of button presses
  solver.minimize(z3.Sum(presses))

  if solver.check() == z3.sat:
    return solver.model().eval(z3.Sum(presses)).as_long()
  return -1

def part2(input):
  total_presses = 0
  for machine in input:
    presses = solve_with_joltage(machine)
    total_presses += presses
  return total_presses


input = open("input/day10.txt", "r").read()
# input = example_input.strip()
print("PART 1 =", part1(parse_input(input)))
print("PART 2 =", part2(parse_input(input)))

File: src/rust/day10.rs
use itertools::Itertools;
use z3::{Optimize, SatResult, ast::Int};

struct Machine {
    lights: Vec<i64>,
    buttons: Vec<Vec<usize>>,
    joltage: Vec<i64>,
}

fn parse_input(input: &str) -> Vec<Machine> {
    let parse_vec = |s: &str| {
        s[1..s.len() - 1]
            .split(',')
            .filter_map(|n| n.parse().ok())
            .collect::<Vec<i64>>()
    };

    input
        .lines()
        .map(|line| {
            let (lights_str, rest) = line.split_once(' ').unwrap();
            let mut parts = rest.split(' ').collect::<Vec<_>>();
            let joltage = parse_vec(parts.pop().unwrap());

            Machine {
                lights: lights_str[1..lights_str.len() - 1]
                    .chars()
                    .map(|c| (c == '#').into())
                    .collect(),
                buttons: parts
                    .iter()
                    .map(|p| parse_vec(p).into_iter().flat_map(usize::try_from).collect())
                    .collect(),
                joltage,
            }
        })
        .collect()
}

fn solve_indicator_lights(machine: &Machine, use_joltage: bool) -> i64 {
    let solver = Optimize::new();
    let zero = Int::from_i64(0);

    let presses = (0..machine.buttons.len())
        .map(|i| {
            let v = Int::new_const(format!("button_{i}"));
            // P[i] > 0
            solver.assert(&v.ge(&zero));
            v
        })
        .collect_vec();

    let targets = if use_joltage { &machine.joltage } else { &machine.lights };

    targets.iter().enumerate().for_each(|(i, &target)| {
        let effective_button_presses = machine
            .buttons
            .iter()
            .zip(presses.iter())
            // Only select P[i] where button affects component i
            .filter(|(b, _)| b.contains(&i))
            .map(|(_, p)| p)
            .collect::<Vec<_>>();
        let equation = if use_joltage {
            // sum(P[i]) = target
            Int::add(&effective_button_presses).eq(Int::from_i64(target))
        } else {
            // Can only be 0 or 1
            // sum(P[i]) % 2 = target
            Int::add(&effective_button_presses)
                .modulo(Int::from_i64(2))
                .eq(Int::from_i64(target))
        };
        solver.assert(&equation);
    });

    // min(sum(P[i]))
    let total_presses = Int::add(&presses);
    solver.minimize(&total_presses);

    if solver.check(&[]) == SatResult::Sat {
        let model = solver.get_model().unwrap();
        model.eval(&total_presses, true).unwrap().as_i64().unwrap()
    } else {
        0
    }
}

pub fn solve_all_indicator_lights(input: &str) -> i64 {
    let machines = parse_input(input);
    machines.iter().map(|m| solve_indicator_lights(m, false)).sum()
}

pub fn solve_all_joltage_requirements(input: &str) -> i64 {
    let machines = parse_input(input);
    machines.iter().map(|m| solve_indicator_lights(m, true)).sum()
}

pub fn main() {
    let input = std::fs::read_to_string("input/day10.txt").expect("Input file not found");
    let now = std::time::Instant::now();
    println!("PART 1 = {}", solve_all_indicator_lights(&input));
    println!("PART 2 = {}", solve_all_joltage_requirements(&input));
    println!("Execution time: {:?}", now.elapsed());
}
